ZIP 316 Revision 1, and corresponding updates to ZIP 320 (#765)
zcash/zips: zip-0316.rst

Privacy impacts of transparent or cross-pool transactions, and the
associated UX issues, will be addressed in ZIP 315 (in preparation).

Specification
=============

Revisions
---------

.. _`Revision 0`:

* Revision 0: The initial version of this specification.

.. _`Revision 1`:

* Revision 1: This version adds support for `MUST-understand Typecodes`_
  and `Address Expiration Metadata`_. It also drops the restriction that
  a UA/UVK must contain at least one shielded Item.

A Revision 0 UA/UVK is distinguished from a Revision 1 UA/UVK by its
Human-Readable Part, as follows.

Let *prefix* be:
* “``u``”, if this is a UA/UVK prior to `Revision 1`_;
* “``ur``”, if this is a UA/UVK from `Revision 1`_ onward.

The Human-Readable Parts (as defined in [#bip-0350]_) of Unified
Addresses are defined as:
* *prefix*, for Unified Addresses on Mainnet;
* *prefix* || “``test``”, for Unified Addresses on Testnet.

> 

Reviewer: ```suggestion
i.e. ``"ur"`` for Mainnet and ``"urtest"`` for Testnet.

```

Reviewer 2: I'm not sure this clarifies anything.

that the MUST-understand specification is correctly enforced even for
such implementations.

.. raw:: html

   </details>

Address Expiration Metadata
---------------------------

As of `Revision 1`_, Typecodes :math:`\mathtt{0xE0}` and :math:`\mathtt{0xE1}`
are reserved for optional address expiry metadata. A producer MAY choose to
generate Unified Addresses containing either or both of the following Metadata
Item Types, or none.

The value of a :math:`\mathtt{0xE0}` item MUST be an unsigned 32-bit integer in
little-endian order specifying the Address Expiry Height, a block height of the
Zcash chain associated with the UA/UVK. A Unified Address containing metadata
Typecode :math:`\mathtt{0xE0}` MUST be considered expired when the height of
the Zcash chain is greater than this value.

The value of a :math:`\mathtt{0xE1}` item MUST be an unsigned 64-bit integer in
little-endian order specifying a Unix Epoch Time, hereafter referred to as the
Address Expiry Time. A Unified Address containing Metadata Typecode
:math:`\mathtt{0xE1}` MUST be considered expired when the current time is
after the Address Expiry Time.

A Sender that supports `Revision 1`_ of this specification MUST set

> a non-zero ``nExpiryHeight`` field in transactions that it creates. If the

Reviewer: ```suggestion
a non-zero ``nExpiryHeight`` field in transactions that it creates if they 
are sending to an address that defines an Address Expiry Height. If the
```

Reviewer 2: (I changed "they are" to "it is" because a Sender is a process acting on behalf of a user, not the user themself. We use this convention generally in Zcash docs.)

``nExpiryHeight`` normally constructed by the Sender would be greater than the
Address Expiry Height, then the transaction MUST NOT be sent. If only an
Address Expiry Time is specified, then the Sender SHOULD choose a value for
``nExpiryHeight`` such that the transaction will expire no more than 24 hours
after the current time. If both :math:`\mathtt{0xE0}` and :math:`\mathtt{0xE1}`
Metadata Items are present, then both restrictions apply.

> If wallet user attempts to send to an expired address, the error presented to

Reviewer: ```suggestion
If a wallet user attempts to send to an expired address, the error presented to
```

little-endian order specifying the Address Expiry Height, a block height of the
Zcash chain associated with the UA/UVK. A Unified Address containing metadata
Typecode :math:`\mathtt{0xE0}` MUST be considered expired when the height of
the Zcash chain is greater than this value.

The value of a :math:`\mathtt{0xE1}` item MUST be an unsigned 64-bit integer in
little-endian order specifying a Unix Epoch Time, hereafter referred to as the
Address Expiry Time. A Unified Address containing Metadata Typecode
:math:`\mathtt{0xE1}` MUST be considered expired when the current time is
after the Address Expiry Time.

A Sender that supports `Revision 1`_ of this specification MUST set
a non-zero ``nExpiryHeight`` field in transactions it creates that are
sent to a Unified Address that defines an Address Expiry Height. If the
``nExpiryHeight`` normally constructed by the Sender would be greater than the
Address Expiry Height, then the transaction MUST NOT be sent. If only an
Address Expiry Time is specified, then the Sender SHOULD choose a value for
``nExpiryHeight`` such that the transaction will expire no more than 24 hours

> after the current time. If both :math:`\mathtt{0xE0}` and :math:`\mathtt{0xE1}`

Reviewer 3: > after the current time

If the address expires 30 *days* from now, why should a transaction expire within 24 hours of when it is sent? Shouldn't we say the transaction should expire no later than 24 hours before the address expiration time?

Deriving a UIVK from a UFVK
---------------------------

As a consequence of the specification in `MUST-understand Typecodes`_,
as of `Revision 1`_ a Consumer of a UFVK MUST understand the meaning of
all "MUST-understand" Metadata Item Typecodes present in the UFVK in
order to derive a UIVK from it.

If the source UFVK is Revision 1 then the derived UIVK SHOULD be

> Revision 1; if it includes any Metadata Items then it MUST be Revision 1.

Reviewer: ```suggestion
Revision 1; if the UIVK includes any Metadata Items then it MUST be Revision 1.
```

In each case, the Typecode remains the same as in the FVK.

Items (including Metadata Items that are not "MUST-understand") that
are unrecognised by a given Consumer, or that are specified in experiments
that the user has not opted into (see `Experimental Usage`_), MUST be

> dropped when deriving a UIVK from a UFVK.

Reviewer: ```suggestion
dropped when a Consumer is deriving a UIVK from a UFVK.
```

Items (including Metadata Items that are not "MUST-understand") that
are unrecognised by a given Consumer, or that are specified in experiments
that the user has not opted into (see `Experimental Usage`_), MUST be

> dropped when when the Consumer derives a UIVK from a UFVK.

Reviewer 2: Duplicate "when". I'll fix it.

In each case, the Typecode remains the same as in the IVK.

Items (including Metadata Items that are not "MUST-understand") that

> are unrecognised by a given Consumer, or that are specified in experiments

Reviewer: Consider re-wording this to say "when the Consumer does not recognise"

that the user has not opted into (see `Experimental Usage`_), MUST be

> dropped when deriving a Unified Address from a UIVK.

Reviewer: ```suggestion
dropped when the Consumer derives a Unified Address from a UIVK.
```
